### ISSUE TITLE
Slicing an interval tree of spans incorrectly excludes spans

### ISSUE THREAD

**Issue by portstrom** [xi-editor/xi-editor#1121]
The documentation of [`tree::NodeInfo::interval`](https://github.com/xi-editor/xi-editor/blob/0297adb8d3ac692ff48ee69cf9e8beda97cdc522/rust/rope/src/tree.rs#L55) says interval trees may need to override this method. The [trait implementation for `spans::SpanInfo`](https://github.com/xi-editor/xi-editor/blob/0297adb8d3ac692ff48ee69cf9e8beda97cdc522/rust/rope/src/spans.rs#L86) doesn't override it. As a result, when slicing an interval tree of spans, any tree node containing only spans that start before the given interval is considered as not intersecting the interval, even if it contains spans that intersect the interval.

Adding this implementation solves the issue:
```rust
    fn interval(&self, len: usize) -> Interval {
        self.iv
    }
```
Test case:
```rust
    let mut builder = SpansBuilder::new(1000);
    builder.add_span(Interval::new(0, 1000), true);
    for position in 1..999 {
        builder.add_span(Interval::new(position, position + 1), false);
    }
    let spans = builder.build().subseq(999..);
    let mut iterator = spans.iter();
    assert_eq!(iterator.next(), Some((Interval::new(0, 1), &true)));
    assert_eq!(iterator.next(), None);
```

**Comment by raphlinus** [comment]
Good catch, it's striking that it hasn't been caught until now. I guess for syntax highlighting most of the spans are short (within a line, generally), so statistically unlikely to trigger this behavior.

Sounds like you're 90% of the way to a PR, would be happy to review one.

**Comment by cmyr** [comment]
This is interesting.

I agree that the logic error is real, and this fix looks correct.

What's curious to me is the actual expected behaviour of overlapping spans; the only way we use the interval tree currently is with the expectation that spans are non-overlapping.

It isn't unreasonable that an interval tree have overlapping spans, but _by convention_ we don't expect this in current use.

I'm not sure what I think about this? It could be helpful to update the docs to include some mention of expected use, at least, and arguably we might even consider encoding this distinction in the type system?

In any case this is orthogonal to the current issue, but I wanted to point it out while I had this state in mind.

**Comment by raphlinus** [comment]
The generic idea of an interval tree has no problem with overlapping spans, and I think if you use spans to indicate markdown-style formatting, that can happen (also if a language server has both syntax and other annotation spans). I think it's just the case that it doesn't happen for syntax highlighting specifically, so at *this* level I don't think we should require or enforce an invariant that spans are non-overlapping.

**Comment by portstrom** [comment]
@cmyr The term [interval tree](https://en.wikipedia.org/wiki/Interval_tree) comes with the assumption that intervals are overlapping. That's what sets it apart from an ordinary tree. Anyway, it would be of course be good to update the documentation to make it clear to those who aren't familiar with all obscure technical terms, as well as add a high level explanation of how it works. It took me some hours to figure out by reading the code.

**Comment by cmyr** [comment]
Sure, I'm more just interested in how the data structure is currently used.

It's an invariant that _style_ spans not overlap; if a style is added that overlaps an existing style, we merge them. If you send the frontend style spans that overlap, it will misbehave.

Should we enforce this at the type level? It would not be hard to make `Spans` be non-overlapping, with merge semantics, and have an `IntervalTree` type with the current behaviour.

@portstrom I agree that a documentation fix would be a good goal for this current discussion.

**Comment by cmyr** [comment]
This was reverted in #1131, due to another issue; this bug is outstanding and should be resolved.
